src/components/node-graph/components/step/node-input.tsx

```diff
@@ -12,21 +12,27 @@ type OwnProps = {
 
 const NodeInput: React.FC<OwnProps> = ({ className = [], value, onChange }) => {
   const debouncedHandleChange = useDebouncedCallback(onChange, 1000);
-  const [displayValue, setDisplayValue] = useState(value);
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    if (inputRef.current) {
+      inputRef.current.value = value;
+    }
+    inputRef.current?.focus();
+  }, [value]);
+
```

Comment:
i don't think the issue is in the implementation of `node-input`, the re-render is caused by `key = {input.id}`

https://github.com/uniconhq/unicon-frontend/blob/042fb872d325406ad90da13862fa32ca7c421357/src/features/tasks/forms/programming-form.tsx#L217-L236

`input.id` changes and it triggers a re-render

_this applies to step metadata component as well_

Reply:
gonna add the commits for this fix in a bit